File: AtCoder/cf17-final/G.cpp
#include <iostream>
#include <vector>

using namespace std;

const int MOD = 1000000007;

int main(){
    int N, K; cin >> N >> K;
    vector<long long> dp(1, 1);
    for(int i=N;i>=1;i--){
        vector<long long> next(dp.size()+(dp.size()-1)/i+1, 0);
        for(int j=0;j<dp.size();j++){
            for(int k=0;k<=K;k++){
                int x = k <= i ? j+(j+k)/i : j;
                next[x] = (next[x] + dp[j]) % MOD;
            }
        }
        dp = next;
    }
    long long res = K*(K+1)/2*N%MOD;
    for(int i=1;i<N;i++) res = (res * (K+1)) % MOD;
    for(int i=1;i<dp.size();i++){
        res = (res + MOD - (i*dp[i])%MOD) % MOD;
    }
    cout << res << endl;
}

File: AtCoder/ABC308/src/bin/g.rs
use proconio::input;
use std::collections::{BTreeMap, BTreeSet};

fn main() {
    input! { q:usize }
    let mut res = BTreeSet::new();
    let mut res_cnt = BTreeMap::new();
    let mut values = BTreeSet::new();
    let mut value_cnt = BTreeMap::new();
    for _ in 0..q {
        input! { c:usize }
        match c {
            1 => {
                input! { x: u32 }
                let left = values.range(x..).next();
                let right = values.range(..x).last();
                if let Some(v) = left {
                    res.insert(v ^ x);
                    *res_cnt.entry(v ^ x).or_insert(0) += 1;
                    if let Some(w) = right {
                        let val = v ^ w;
                        *res_cnt.entry(val).or_insert(0) -= 1;
                        if res_cnt[&val] == 0 {
                            res.remove(&val);
                            res_cnt.remove(&val);
                        }
                    }
                }
                if let Some(v) = right {
                    res.insert(v ^ x);
                    *res_cnt.entry(v ^ x).or_insert(0) += 1;
                }
                values.insert(x);
                *value_cnt.entry(x).or_insert(0) += 1;
            }
            2 => {
                input! { x: u32 }
                *value_cnt.entry(x).or_insert(0) -= 1;
                if value_cnt[&x] == 0 {
                    values.remove(&x);
                    value_cnt.remove(&x);
                }
                let left = values.range(x..).next();
                let right = values.range(..x).last();
                if let Some(v) = left {
                    let val = v ^ x;
                    *res_cnt.entry(val).or_insert(0) -= 1;
                    if res_cnt[&val] == 0 {
                        res.remove(&val);
                        res_cnt.remove(&val);
                    }
                    if let Some(w) = right {
                        res.insert(v ^ w);
                        *res_cnt.entry(v ^ w).or_insert(0) += 1;
                    }
                }
                if let Some(v) = right {
                    let val = v ^ x;
                    *res_cnt.entry(val).or_insert(0) -= 1;
                    if res_cnt[&val] == 0 {
                        res.remove(&val);
                        res_cnt.remove(&val);
                    }
                }
            }
            _ => {
                println!("{}", res.iter().next().unwrap());
            }
        }
    }
}
